api/authentication.go

```diff
@@ -4,127 +4,145 @@ import (
-		token, err := h.getAuthTokenFromRequest(req)
-		if err != nil {
-			h.Error(resp, http.StatusUnauthorized, err)
-			return
-		}
-		userId, err := h.getUserIdFromToken(token)
-		if err != nil {
-			h.Error(resp, http.StatusUnauthorized, err)
-			return
-		}
+func (h apiHandler) checkScopes(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		scopes, ok := r.Context().Value(BearerScopes).([]string)
+		if ok {
+			if err := h.isAuthenticated(r); err != nil {
+				h.Error(w, r, http.StatusUnauthorized, err)
+				return
+			}
 
-		user, err := h.verifyUserExists(userId)
-		if err != nil {
-			if err == db.ErrNotFound {
-				h.Error(resp, http.StatusUnauthorized, errors.New("invalid user"))
-			} else {
-				h.Error(resp, http.StatusInternalServerError, err)
+			for _, scope := range scopes {
+				if err := h.verifyUserHasScope(r, scope); err != nil {
+					err := fmt.Errorf("endpoint '%s' requires '%s' scope: %w", r.URL.Path, scope, err)
+					h.Error(w, r, http.StatusForbidden, err)
+					return
+				}
 			}
-			return
 		}
 
-		// Add the user's ID and access level to the list of params
-		ctx := req.Context()
-		ctx = context.WithValue(ctx, currentUserIdCtxKey, user.Id)
-		ctx = context.WithValue(ctx, currentUserAccessLevelCtxKey, user.AccessLevel)
-
-		next.ServeHTTP(resp, req.WithContext(ctx))
-	})
+		next.ServeHTTP(w, r)
+	}
 }
 
-func (h apiHandler) requireAdmin(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
-		if err := h.verifyUserIsAdmin(req); err != nil {
-			h.Error(resp, http.StatusForbidden, err)
-			return
+func (h apiHandler) isAuthenticated(r *http.Request) error {
+	token, err := h.getAuthTokenFromRequest(r)
+	if err != nil {
+		return err
+	}
+
+	claims := token.Claims.(*gompClaims)
+	if len(claims.Scopes) == 0 {
+		return errors.New("token had no scopes")
+	}
+
+	userId, err := h.getUserIdFromToken(token)
+	if err != nil {
+		return err
+	}
+
+	user, err := h.verifyUserExists(userId)
+	if err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			return errors.New("invalid user")
 		}
 
-		next.ServeHTTP(resp, req)
-	})
+		return err
+	}
+
+	// Add the user's ID and token to the list of params
+	ctx := r.Context()
+	ctx = context.WithValue(ctx, currentUserIdCtxKey, user.Id)
+	ctx = context.WithValue(ctx, currentUserTokenCtxKey, token)
+	*r = *r.WithContext(ctx)
+
+	return nil
 }
 
-func (h apiHandler) requireAdminUnlessSelf(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
-		urlId, err := getUserIdFromUrl(req)
-		if err != nil {
-			h.Error(resp, http.StatusBadRequest, err)
-			return
-		}
-		ctxId, err := getResourceIdFromCtx(req, currentUserIdCtxKey)
-		if err != nil {
-			h.Error(resp, http.StatusUnauthorized, err)
-			return
-		}
+func (apiHandler) verifyUserHasScope(r *http.Request, scope string) error {
```

Comment:
A consequence of this approach is that of an admin changes the access level of a user logged in from another machine, that user will continue with their previous access rights.